test/blackbox-tests/test-cases/pkg/disjunction.t

```diff
@@ -0,0 +1,51 @@
+Exercise the behaviour of the solver when packages contain disjunctions in
+their dependencies.
+
+  $ . ./helpers.sh
+  $ mkrepo
+
+Create a package that depends on a disjunction of 3 other packages.
+  $ mkpkg a 0.1
+  $ mkpkg b 0.1
+  $ mkpkg c 0.1
+  $ mkpkg dep-on-disjunction 0.1 << EOF
+  > depends: [ "a" | "b" | "c" ]
+  > EOF
+
+The first package in the disjunction is chosen in this case.
+  $ solve dep-on-disjunction
+  Solution for dune.lock:
+  - a.0.1
+  - dep-on-disjunction.0.1
+
+Release a new version of the second package in the disjunction to demonstrate
+that relative version numbers don't affect the choice.
+  $ mkpkg b 0.2
+  $ solve dep-on-disjunction
+  Solution for dune.lock:
+  - a.0.1
+  - dep-on-disjunction.0.1
+
+New version of dep-on-disjunction specifying version numers of deps. Note that
+only "b" exists with the specified version. If the solver chooses the new
+version of deps-on-disjunction then it must also choose "b" rather than "a".
+  $ mkpkg dep-on-disjunction 0.2 << EOF
+  > depends: [ "a" {= "0.2" } | "b" {= "0.2" } | "c" {= "0.2" } ]
+  > EOF
+  $ solve dep-on-disjunction
+  Solution for dune.lock:
+  - b.0.2
+  - dep-on-disjunction.0.2
+
+Mark "b.0.2" as avoid-version so the solver will try to find a solution which doesn't include it.
+  $ mkpkg b 0.2 << EOF
+  > flags: [avoid-version]
+  > EOF
```

Comment:
To me this seems correct, as `b.0.2` is the only version that can be picked, even if it is to be avoided otherwise.

In opam-monorepo we reshuffle the candidates to demote avoid-version packages: https://github.com/tarides/opam-monorepo/blob/141d351b79715c99d66bfd898a2a64115a2e8769/lib/opam_solve.ml#L179-L189 

This is of course only a partial solution because if all versions are demoted, but the package is first in the disjunction, the solver will still pick it over the next one in the disjunction (even if these packages are not marked as `avoid-version`).

Reply:
That's also the approach I took with this fix https://github.com/ocaml/dune/pull/10596